packages/realm_dart/lib/src/realm_class.dart

```diff
@@ -388,10 +392,10 @@ class Realm {
   /// Executes the provided [writeCallback] in a temporary write transaction. Both acquiring the write
   /// lock and committing the transaction will be done asynchronously.
   Future<T> writeAsync<T>(T Function() writeCallback, [CancellationToken? cancellationToken]) async {
+    assert(!_isFuture<T>(), 'writeCallback must be synchronous');
     final transaction = await beginWriteAsync(cancellationToken);
-
     try {
-      T result = writeCallback();
+      T result = await writeCallback(); // ignore: await_only_futures
```

Comment:
Why do we await here if we assert the callback isn't a future?

Reply:
😄 ever vigilant. I could just drop it. As is, all that is needed to support async callbacks is ton change the return type of `writeCallback` from `T` to `FutureOr<T>` .. and re-enable some skipped tests. 

I should probably have put this PR back into in-progress

Reply:
.. but it is not a bug per-se .. in dart you can `await` non-futures, even `null` just fine. It differs from .NET on that.

Reply:
I have removed all the new tests again and simply disallowed async callbacks even to `writeAsync`.